### ISSUE TITLE
Realtime change feeds?

### ISSUE THREAD

**Issue by nkev** [scylladb/scylla#611]
As far as I can see, Cassandra triggers are not supported. This is good in my opinion because triggers are usually transactional (slow) and work only on CRUD operations, so I find them primitive and limited. I would much rather see Scylla support a more modern data change notification mechanism, like [RethinkDB change feeds](https://rethinkdb.com/docs/changefeeds/javascript/). RethinkDB is great in this respect, but I find the ReQL querying mechanism very frustrating to use and harder to read compared to SQL. I don't know of a SQL based change feed system out there.

I would love to see SQL syntax like:

``` SQL
SELECT * FROM scores ORDER BY points DESC LIMIT 5 SUBSCRIBE
```

This would return the top 5 scores the first time it is called, but it would also continue returning results to the caller whenever the result of this query changes (via the `SUBSCRIBE` keyword). How cool would that be?

To unsubscribe from this notification, you could later send:

``` SQL
SELECT * FROM scores ORDER BY points DESC LIMIT 5 UNSUBSCRIBE
```

This would be a perfect fit for Scylla's performance oriented architecture to create real-time apps.

If this is not possible due to driver constraints or whatever, a simple way to receive JSON updates to an API should be easier to implement:

``` SQL
SELECT * FROM scores ORDER BY points DESC LIMIT 5 SUBSCRIBE JSON  `http://localhost:8000/update-score`
```

If Cassandra compatibility is paramount, perhaps you could use the `Trigger` command to incorporate the URL somehow.

Thoughts?

**Comment by avikivity** [comment]
This is indeed something we've planned.  But for now we're focusing on
completing cql support, so this will be done after GA, unless you implement
it yourself.

**Comment by nkev** [comment]
I understand. I would certainly implement it if I knew C++ :) I have a C# background and am transitioning to Go. If anyone else wants to implement it, I can certainly be involved in design/testing.

Is there a link to any documentation about your plan on this feature?

**Comment by avikivity** [comment]
No, we haven't started it yet.  You could say the most complete documentation is in this bug report.

**Comment by nkev** [comment]
Ok. Can we please leave this open with an `enhancement` tag, in case someone want to take it up?

**Comment by nkev** [comment]
Is there no C++ guru to take up this challenge?

**Comment by Fornax96** [comment]
I bet a lot has changed in the 5 years since this issue was opened. Is there any way to implement something like this currently?

**Comment by avikivity** [comment]
https://docs.scylladb.com/using-scylla/cdc/

**Comment by Fornax96** [comment]
@avikivity As far as I can tell that's only for viewing row history, right? It doesn't notify us of changes in real time.

**Comment by denesb** [comment]
@Fornax96 yes it is poll-based, not push based.

**Comment by avikivity** [comment]
@Fornax96 we're adding long-poll style APIs to make polling more efficient.

**Comment by seeekr** [comment]
@avikivity is there any work being done currently on implementing more efficient polls and/or even implementing something push-based? It would be really nice to have efficient (near-)real-time CDC. If not, has the latter ever been discussed in any detail? Who might be some good people to cc if I wanted to start an initial design discussion about such a feature?

It would also be very helpful if the docs were quite upfront about the implementation currently being poll-based. That would immediately make it clear that CDC cannot be used for certain features, mainly if efficient low-latency updates are required. Should I send a doc PR?

**Comment by avikivity** [comment]
@kbr- @piodul 

I don't think there is active work here, but the cdc people will know better.

**Comment by kbr-** [comment]
The current implementation of CDC is heavily limited by the data model of Scylla. In this model there is no order of writes, instead writes have timestamps that are used which data is "last". It may happen (and usually does) that Scylla makes a write with timestamp 42, but then a new write appears with timestamp 40.

CDC is supposed to give you an order such that data appearing later in the stream overwrites the data appearing earlier, so we artificially come up with an order for those writes based on the timestamps. To protect ourselves from such "out-of-order" writes, we need to query with a delay (so called confidence window), and push-based solutions are basically out of question.

The topic of real-time and/or push-based CDC will come back when we implement Raft-based strongly consistent tables. There, writes will be given a total order, and we can show a write in CDC as soon as it's committed to the Raft log (it's not possible for a new write to come "earlier").

**Comment by seeekr** [comment]
@kbr- thanks for elucidating your current thought process & design, very helpful. A few remarks:

For some use cases the problem of out-of-order writes is not really a problem. The application may be structured in such a way that concurrent writes to the same data is never going to happen by design. Or the application may be perfectly OK with resolving / reconciling these writes in the CDC consumer or further downstream.

To me, what I perceive as the "essence of Scylla", namely being a very fast data store that gives its users the ability to make deliberate tradeoffs between performance and consistency, would be best expressed in a CDC feature set that would allow me to make that tradeoff myself, not force me into a tradeoff towards more consistency that I don't really need. Yes, if then it's also possible to opt into a consistent view, similar to what LWT offers, with its great ease of use, and with how simple to understand it is, that is excellent. But please don't make that choice for me as a user.

So, from my perspective, what I would have expected in terms of the current design of CDC, is the following: I get exactly the stream of write events that the coordinator for each given write operation is seeing. (I think I have some inaccuracy of my understanding of Scylla's model here. But this is roughly how I thought I heard some docs describe the CDC stream to me, and it made a lot of sense.) Which is exactly the same thing as I would get if I were to just manually poll the CDC table, but I get it in the most efficient fashion it can be done, without incurring any additional reads, unless I specifically ask for some ease-of-use feature to make those for me. (Like with LWTs or with pre-/postimage CDC params etc.)

Does this make sense to you? What do you think?

**Comment by seeekr** [comment]
Something to add: My mental model was more specifically that the coordinator node would just "post its world view into the CDC stream", so to speak, without any additional reads or pretty much any other work that would need to be done. After all, it just completed the client's request, and was able to tell the client that the change it requested has been successfully accomplished. And this is exactly the bit of information that I am most interested in receiving in a different format -- as a stream, just as in real-time as the client gets it, with no further processing done on it.
If a concurrent write should be seen "later" (or even sooner), that can be my application's problem to deal with. At least I should be able to make that decision, if that's already enough for my particular use case. And it seems that from a technical perspective, all that would be needed to make such a stream work, is just adding a scalable (enough) pub/sub mechanism there and streaming the events the coordinators see into that mechanism and on to those "write stream clients" that have requested this particular subset of it. If that makes sense.

**Comment by kbr-** [comment]
> The application may be structured in such a way that concurrent writes to the same data is never going to happen by design.

You can get a failed write, say, due to a timeout. Once you get a failed write, the write can in theory happen at any point in the future. If you start a new write after that failed write, the two can appear in any order. So if you wanted to ensure that his indeed never really happens, you would have to stop writing to a given piece of data as soon as you get the first failed write. Other than that, you could do stuff like resuming writing only after 5 seconds or something, but that gives you at most a probabilistic guarantee, not a complete guarantee that the old write won't appear later.

> I get exactly the stream of write events that the coordinator for each given write operation is seeing.

I guess you could open a channel from a node, and once the node coordinates a write, it would push the write through the channel. You would get writes as they come, and you would get writes with out-of-order timestamps.

Then there's the issue of multiple coordinators. You would need a channel to every node in the system, including remote DCs, because any node in Scylla can coordinate a write to any partition.

Then there's the issue of coordinator failure. What if a coordinator successfully writes, then fails before pushing the write to you? Where do you get that write? Assuming there's a CDC log table, I guess you could have some complex recovery procedure where you poll the CDC log table. But then you have to keep doing this as long as the coordinator is dead - because it might not really be dead, it may just be a network partition. You can only be sure that you receive all pushed messages when all nodes are alive and pinging you.

**Comment by tzach** [comment]
@ultrabug , you had  a similar idea  (rethinkdb like subscribe)

**Comment by ultrabug** [comment]
Indeed @tzach and I :popcorn: reading through the conversation.

RethinkDB had stronger promises than just feeds since it could also propagate calculations like aggregates tho ;)

While I do relate with @seeekr needs, I have to see this feature request from the Scylla folks standpoint. Development energy is not infinite so I'd rather see stable Raft powered Scylla than this feature in the short term.

**BUT** I'm warning you @tzach, when Raft is here, I want that realtime change feature alright :laughing:
